dask_cuda/device_host_file.py

```diff
@@ -82,19 +82,26 @@ def _(header, frames):
     return DeviceSerialized(header["main-header"], parts, header["is-cuda"])
 
 
+def copy_to_host(ary):
+    if hasattr(ary, "copy_to_host"):
+        return ary.copy_to_host()
+    else:
+        return cuda.as_cuda_array(ary).copy_to_host()
+
+
 def device_to_host(obj: object) -> DeviceSerialized:
     header, frames = serialize(obj, serializers=["cuda", "pickle"])
     is_cuda = [hasattr(f, "__cuda_array_interface__") for f in frames]
     frames = [
-        cuda.as_cuda_array(f).copy_to_host() if ic else f
+        copy_to_host(f) if ic else f
         for ic, f in zip(is_cuda, frames)
     ]
     return DeviceSerialized(header, frames, is_cuda)
 
 
 def host_to_device(s: DeviceSerialized) -> object:
     frames = [
-        cuda_memory_manager.to_device(f) if ic else f
+        cuda_memory_manager.to_device(f.ravel().view("u1")) if ic else f
```

Comment:
I don't think I understand, why is this necessary? Also, is it necessary for all cases or just specific ones?

Reply:
Because RMM `DeviceBuffer`'s require [1-D contiguous data that is `uint8`]( https://github.com/rapidsai/rmm/blob/1002230ce362dce1842f5362b1814ce247889fc5/python/rmm/_lib/device_buffer.pyx#L131 ) since they don't handle size or type.

That said, Dask serializers are able to use the header to reconstruct the shape and type given a buffer like this. For example this can be seen in [CuPy]( https://github.com/dask/distributed/blob/9af811d8f9858c63b9586bcfb78ce2dec8f5d6b3/distributed/protocol/cupy.py#L55 ) and [Numba]( https://github.com/dask/distributed/blob/9af811d8f9858c63b9586bcfb78ce2dec8f5d6b3/distributed/protocol/numba.py#L34 ). cuDF itself always works with [1-D untyped `Buffer`]( https://github.com/rapidsai/cudf/blob/ba5aaeb44a1acdb00728db138e9c82014caa92a4/python/cudf/cudf/core/buffer.py#L10-L11 ) objects [under-the-hood]( https://github.com/rapidsai/cudf/blob/a0673be1eebc7a1746bf04655648cf97aaf6c47f/python/cudf/cudf/core/column/column.py#L36-L44 ). Here's [one example of cuDF's serialization]( https://github.com/rapidsai/cudf/blob/ba5aaeb44a1acdb00728db138e9c82014caa92a4/python/cudf/cudf/core/column/column.py#L877 ). There are others that handle various specific types, but all more-or-less work the same.

Reply:
Wow, you didn't need to provide all those examples, the first sentence would have sufficed. But thanks for the long description anyway!

Reply:
Haha ok. 😄 Yeah I figured that was just useful info to share after having done the homework on it earlier 😉